include/lock/optimistic_lock.hpp

```diff
@@ -199,6 +228,35 @@ class OptimisticLock
     }
   }
 
+  /**
+   * @brief Get a shared lock with an intent-exclusive lock if version is same.
+   *
+   * @param ver an expected version value (turned off by kSIXAndSBitsMask).
+   * @retval true if the given version value is the same as a current one.
+   * @retval false otherwise.
+   */
+  [[nodiscard]] auto
+  TryLockSIX(const uint64_t ver)  //
+      -> bool
+  {
+    auto expected = ver | (lock_.load(std::memory_order_relaxed) & ~kSBitsMask);
+    const auto desired = expected | kSIXLock;
+    while (true) {
+      for (size_t i = 1; true; ++i) {
+        const auto cas_success = lock_.compare_exchange_weak(
+            expected, desired, std::memory_order_acquire, std::memory_order_relaxed);
+        if (cas_success) return true;
```

Comment:
実はここはフェンスなしでいけます（たぶん）．SIXロックを取ることでこれ以降も他のスレッドがロック領域を変更しないことが保証されるので，このCASの実行順が前後しても読む値は変わりません（たぶん）．

```suggestion
        if (lock_.compare_exchange_weak(expected, desired, std::memory_order_relaxed)) return true;
```